ZEPPELIN-210 Show notebook in a homescreen

Leemoonsoo opened the pull request:
https://issues.apache.org/jira/browse/ZEPPELIN-210

This patch lets user specify a notebook id by `ZEPPELIN_NOTEBOOK_HOMESCREEN` environment variable or `zeppelin.notebook.homescreen` property. Then Zeppelin default homescreen will be replaced by the notebook.

Here's an example of home screen
![image](https://cloud.githubusercontent.com/assets/1540981/9153389/8039610e-3e07-11e5-8ef5-5407798d86b1.png)

Leemoonsoo commented:
Ready to merge.

eranwitkon commented:
I tested this feature but could not get it to work.
1) we should add the property place holder to zeppelin-site.xml.template and env variable to zeppelin-env.sh.template
```
<property>
  <name>zeppelin.notebook.homescreen</name>
  <value>2AV4WUEMK/</value>
</property>
```
```
export ZEPPELIN_NOTEBOOK_HOMESCREEN=2AV4WUEMK
```
2) Adding the property to site XML does not work, only the env variable works.
3) note that if we rename the notebook we still have to set the variable to the id (file name on the disk) and not the notebook name. (only why to find the id is from the URL)
4) Can we hide the homepage notebook from the notebook list? I wouldn't want all users play with the homepage...

Eran

Leemoonsoo commented:
@eranwitkon  Thanks for feedback.


> 1) we should add the property place holder to zeppelin-site.xml.template and env variable to zeppelin-env.sh.template

Added property place holder to zeppelin-site.xml.template and zeppelin-env.sh.template


> 2) Adding the property to site XML does not work, only the env variable works.

To me both way works. could you verify it again?

> 3) note that if we rename the notebook we still have to set the variable to the id (file name on the disk) and not the notebook name. (only why to find the id is from the URL)

For now, user need to get notebook id from URL. not name of notebook.

> 4) Can we hide the homepage notebook from the notebook list? I wouldn't want all users play with the homepage...

Added option to selectively hide homescreen notebook from list.

eranwitkon commented:
Hi,
I tested this again with the following results:
1) I can set the variable both in the *-env and *-site and get it to work
2) hide the notebook from the list works fine
BUT I had a problem with my MD notebook when I set the tutorial notebook as
home page ( 2A94M5J1Z ) everything worked fine but setting my md notebook
showed an empty page.
attached is my md notebook
Here is my homepage open as notebook:
[image: Selection_010.png]
Here is what I get when I set it as homepage
[image: Selection_011.png]
3) idea: can we add a button on the notebook level by the schedule button
with the "Set as homepage" option? instead of editing the env\site files?
4) idea: out of the scope of this PR but should we have a zeppelin function
which list the notebooks in zeppelin e.g. z.notebooks collection, so I can
write my personalized homepage but still get the list of notebooks? - if
needed I can see the current homepage implementation and refactor it as a z
method.
Eran

On Mon, Aug 10, 2015 at 1:03 AM Lee moon soo <notifications@github.com>
wrote:

> @eranwitkon <https://github.com/eranwitkon> Thanks for feedback.
>
> 1) we should add the property place holder to zeppelin-site.xml.template
> and env variable to zeppelin-env.sh.template
>
> Added property place holder to zeppelin-site.xml.template and
> zeppelin-env.sh.template
>
> 2) Adding the property to site XML does not work, only the env variable
> works.
>
> To me both way works. could you verify it again?
>
> 3) note that if we rename the notebook we still have to set the variable
> to the id (file name on the disk) and not the notebook name. (only why to
> find the id is from the URL)
>
> For now, user need to get notebook id from URL. not name of notebook.
>
> 4) Can we hide the homepage notebook from the notebook list? I wouldn't
> want all users play with the homepage...
>
> Added option to selectively hide homescreen notebook from list.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-129253727>
> .
>

anthonycorbacho commented:
What happen if I set a note as Home and then remove this note?

eranwitkon commented:
based on my test it revert back to the original homepage.
Eran

On Mon, Aug 10, 2015 at 9:13 AM Anthony Corbacho <notifications@github.com>
wrote:

> What happen if I set a note as Home and then remove this note?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-129318414>
> .
>

Leemoonsoo commented:
@eranwitkon Thanks for the feedback.

> 2) hide the notebook from the list works fine
> BUT I had a problem with my MD notebook when I set the tutorial notebook as
> home page ( 2A94M5J1Z ) everything worked fine but setting my md notebook
> showed an empty page.
> attached is my md notebook
> Here is my homepage open as notebook:
> [image: Selection_010.png]
> Here is what I get when I set it as homepage
> [image: Selection_011.png]

Could you re-attache screenshot?

> 3) idea: can we add a button on the notebook level by the schedule button
> with the "Set as homepage" option? instead of editing the env\site files?

If there's "Set as homepage" button, the information need to be saved.
I think note.json is not the place, because two or more notebook can have "set as homepage" option on, while note.json can be copied from the other instance of Zeppelin.
Information need to be saved in somewhere Zeppelin instance specific place (like conf/interpreter.json).

I think having "set as homepage" button is big enough to discuss and implement separately.

> 4) idea: out of the scope of this PR but should we have a zeppelin function
> which list the notebooks in zeppelin e.g. z.notebooks collection, so I can
> write my personalized homepage but still get the list of notebooks? - if
> needed I can see the current homepage implementation and refactor it as a z
> method.

This is good idea and can be a future work

eranwitkon commented:
Here they are re-attached

On Mon, Aug 10, 2015 at 10:23 PM Lee moon soo <notifications@github.com>
wrote:

> @eranwitkon <https://github.com/eranwitkon> Thanks for the feedback.
>
> 2) hide the notebook from the list works fine
> BUT I had a problem with my MD notebook when I set the tutorial notebook as
> home page ( 2A94M5J1Z ) everything worked fine but setting my md notebook
> showed an empty page.
> attached is my md notebook
> Here is my homepage open as notebook:
> [image: Selection_010.png]
> Here is what I get when I set it as homepage
> [image: Selection_011.png]
>
> Could you re-attache screenshot?
>
> 3) idea: can we add a button on the notebook level by the schedule button
> with the "Set as homepage" option? instead of editing the env\site files?
>
> If there's "Set as homepage" button, the information need to be saved.
> I think note.json is not the place, because two or more notebook can have
> "set as homepage" option on, while note.json can be copied from the other
> instance of Zeppelin.
> Information need to be saved in somewhere Zeppelin instance specific place
> (like conf/interpreter.json).
>
> I think having "set as homepage" button is big enough to discuss and
> implement separately.
>
> 4) idea: out of the scope of this PR but should we have a zeppelin function
> which list the notebooks in zeppelin e.g. z.notebooks collection, so I can
> write my personalized homepage but still get the list of notebooks? - if
> needed I can see the current homepage implementation and refactor it as a z
> method.
>
> This is good idea and can be a future work
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-129572986>
> .
>

Leemoonsoo commented:
@eranwitkon i still can not see re-attached screenshots.

eranwitkon commented:
Here is what I see when I set a a spesific notebook as homepage
![selection_011](https://cloud.githubusercontent.com/assets/2227083/9216948/e854f766-40c9-11e5-9db9-f246b11197b0.png)

here is the JSON code for the notebook I used (can't attach the file)
```
{
  "paragraphs": [
    {
      "text": "%md\nThis is my homepage",
      "config": {
        "colWidth": 6.0,
        "graph": {
          "mode": "table",
          "height": 300.0,
          "optionOpen": false,
          "keys": [],
          "values": [],
          "groups": [],
          "scatter": {}
        },
        "editorHide": false,
        "tableHide": true
      },
      "settings": {
        "params": {},
        "forms": {}
      },
      "jobName": "paragraph_1439146002471_-193197095",
      "id": "20150809-214642_1580883980",
      "result": {
        "code": "SUCCESS",
        "type": "HTML",
        "msg": "\u003cp\u003eThis is my homepage\u003c/p\u003e\n"
      },
      "dateCreated": "Aug 9, 2015 9:46:42 PM",
      "dateStarted": "Aug 9, 2015 9:47:22 PM",
      "dateFinished": "Aug 9, 2015 9:47:23 PM",
      "status": "FINISHED",
      "progressUpdateIntervalMs": 500
    },
    {
      "config": {
        "colWidth": 12.0,
        "graph": {
          "mode": "table",
          "height": 300.0,
          "optionOpen": false,
          "keys": [],
          "values": [],
          "groups": [],
          "scatter": {}
        },
        "editorHide": false,
        "tableHide": true
      },
      "settings": {
        "params": {},
        "forms": {}
      },
      "jobName": "paragraph_1439146042936_-147350679",
      "id": "20150809-214722_1019915871",
      "dateCreated": "Aug 9, 2015 9:47:22 PM",
      "status": "READY",
      "progressUpdateIntervalMs": 500
    }
  ],
  "name": "Homepage",
  "id": "2AV4WUEMK",
  "angularObjects": {
    "2AUFAQ9QP": [],
    "2AXFQ16AT": [],
    "2AWAG511N": [],
    "2AUNC5A3V": [],
    "2AV3F1EXF": [],
    "2AXWVZ4V2": [],
    "2AXE167H3": [],
    "2AUUVVG1K": [],
    "2AWYKXSWN": [],
    "2AWESBBRD": [],
    "2AUYYWTQ3": []
  },
  "config": {
    "looknfeel": "default"
  },
  "info": {}
}
```

Other notebooks works fine.

Leemoonsoo commented:
@eranwitkon I tested with your notebook. It's because of Homescreen is only displaying results. Once you run your markdown paragraph and get result, you'll see it in Homescreen.

eranwitkon commented:
I would expect that if we set a page as home page we need to execute it
automaticly on load, right?

בתאריך יום ה׳, 13 באוג׳ 2015, 19:39 מאת Lee moon soo <
notifications@github.com>:

> @eranwitkon <https://github.com/eranwitkon> I tested with your notebook.
> It's because of Homescreen is only displaying results. Once you run your
> markdown paragraph and get result, you'll see it in Homescreen.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-130753541>
> .
>

Leemoonsoo commented:
Some user might not want to automatic execution for his homescreen notebook. Check button like "Run this notebook on startup" for all notebooks would give better flexibility.

eranwitkon commented:
Yes, do we dupport this feature? 'Run at startup?'

בתאריך יום ה׳, 13 באוג׳ 2015, 19:54 מאת Lee moon soo <
notifications@github.com>:

> Some user might not want to automatic execution for his homescreen
> notebook. Check button like "Run this notebook on startup" for all
> notebooks would give better flexibility.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-130759416>
> .
>

Leemoonsoo commented:
Not yet, but i think the feature could be useful.

eranwitkon commented:
I agree.
I see home page in report mode as the most common use case (right?) So any
user using it in report mode need to run it in startup, (right?)
Eran

בתאריך יום ה׳, 13 באוג׳ 2015, 20:04 מאת Lee moon soo <
notifications@github.com>:

> Not yet, but i think the feature could be useful.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-130762140>
> .
>

eranwitkon commented:
Where are we going to document this option? (as it is a user option and not a developer option?
I would like to propose adding this part to the documentation:

## Adding Notebooks list to your custom Homepage
If you want to have your custom homepage but still be able to display the list of notebooks on your home page all you need to do is use our %angular support.
Add the following code to a paragraph in you home page and run it... walla! you have your notebooks list.
```javascript
println("""%angular 
        <div class="col-md-4" ng-controller="HomeCtrl as home">
        <h4>Notebooks</h4>

        <div>
           <h5><a href="" data-toggle="modal" data-target="#noteNameModal" style="text-decoration: none;">
              <i style="font-size: 15px;" class="icon-notebook"></i> Create new note</a></h5>
          <ul style="list-style-type: none;">
            <li ng-repeat="note in home.notes.list track by $index"><i style="font-size: 10px;" class="icon-doc"></i>
              <a style="text-decoration: none;" href="#/notebook/{{note.id}}">{{note.name || 'Note ' + note.id}}</a>
            </li>
          </ul>
        </div>
      </div>
      """)
```
Once you run it you will see output similar to this one:
![selection_012](https://cloud.githubusercontent.com/assets/2227083/9259047/77dedd32-4207-11e5-9013-44d53565aca0.png)

And in report mode
![selection_013](https://cloud.githubusercontent.com/assets/2227083/9259094/9ed419ca-4207-11e5-935a-aae335f009c1.png)

Main trick here relays in linking the ```<div>``` to the controller:
```javascript
<div class="col-md-4" ng-controller="HomeCtrl as home">
```
once we have ```home``` as our controller variable in our ```<div></div>``` we can use ```home.notes.list``` to get access to the notebook list.

Leemoonsoo commented:
@eranwitkon  Thanks for the nice tip! How about creating one sub page in "Manual" section, Above / Below "Dynamic Form" page in http://zeppelin.incubator.apache.org/docs/index.html?

eranwitkon commented:
Added the documentation - see my pull request
https://github.com/apache/incubator-zeppelin/pull/214
Eran

On Fri, Aug 14, 2015 at 5:38 PM Lee moon soo <notifications@github.com>
wrote:

> @eranwitkon <https://github.com/eranwitkon> Thanks for the nice tip! How
> about creating one sub page in "Manual" section, Above / Below "Dynamic
> Form" page in http://zeppelin.incubator.apache.org/docs/index.html?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/191#issuecomment-131128187>
> .
>
